src/Voron/Impl/FreeSpace/StreamBitArray.cs

```diff
@@ -84,88 +86,177 @@ private void Write(byte* ptr)
+            var alignedPos = i * 32 + alignedBits;
+            var unalignedPos = i * 32 + unalignedBits + 16;
+            if (unalignedPos + num > CountOfItems * 32)
+                hasUnaligned = false;
+
+            switch (hasAligned, hasUnaligned)
+            {
+                case (true, true):
+                    return Math.Min(alignedPos, unalignedPos);
+                case (true, false):
+                    return alignedPos;
+                case (false, true):
+                    return unalignedPos;
+            }
+        }
+        
+        return -1;
     }
 
-    private interface IModifyBuffer
+    private int FindRangeUsingWords(int num)
     {
-        uint Modify(uint i);
-        Vector256<uint> Modify(Vector256<uint> v);
+        Debug.Assert(num >= 32, "num >= 32");
+        int requiredWords = num / 32;
+        int totalWords = requiredWords + (num % 32 > 0 ? 1 : 0);
+            
+        for (int i = 0; i < CountOfItems; i += Vector256<uint>.Count)
+        {
+            var a = Vector256.LoadUnsafe(ref _inner[i]);
+            var eq = Vector256.Equals(a, Vector256<uint>.AllBitsSet);
+            if (eq == Vector256<uint>.Zero)
+                continue;
+            
+            var mask = eq.ExtractMostSignificantBits();
+            int idx = BitOperations.TrailingZeroCount(mask) + i;
+            
+            if (idx + totalWords > CountOfItems)
+                break;
+
+            while(true)
+            {
+                Debug.Assert(_inner[idx] == uint.MaxValue, "_inner[idx] == uint.MaxValue - because it must be given the SIMD check");
+                int remainingBits = num % 32;
+                int found = idx * 32;
+                if (idx > 0)
+                {
+                    var prev = _inner[idx - 1];
+                    int previousBits = BitOperations.LeadingZeroCount(~prev);
+                    if (previousBits > 0)
+                    {
+                        int bitsToTake = Math.Min(remainingBits, previousBits);
+                        remainingBits -= bitsToTake;
+                        found -= bitsToTake;
+                    }
+                }
+                
+                for (int j = 1; j < requiredWords; j++)
+                {
+                    if (_inner[idx + j] == uint.MaxValue) 
+                        continue;
+
+                    idx += j;
+                    goto NextElement;
+                }
+
+                if (remainingBits == 0)
+                    return found;
+
+                var availableBits = BitOperations.TrailingZeroCount(~_inner[idx + requiredWords]);
+                if(availableBits >= remainingBits)
+                    return found;
+
+                idx += requiredWords;
+                    
+                NextElement:
+                while (idx < i + Vector256<uint>.Count && 
+                       _inner[idx] != uint.MaxValue)
+                {
```

Comment:
Used like this, there is no way to elide the `_inner` bound checks.

Reply:
`_inner` isn't an array, there aren't any bound checks here, see:

https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIGYACAVwDtdsAzGB3DKJsBgYBhBgG8aDKQ0nT6DDgEsEMACYNFLIQEkMMfLgDaADgC6AbllSrDeZqEBZABT2G2AJTib02wHYGuvpG2BY2AL40YUA==